core/protocol/all_services.go

```diff
@@ -330,40 +331,35 @@ func newServices(
 	if svcs.spam != nil {
 		svcs.snapshot.AddProviders(svcs.spam)
 	}
-	powWatchers := []netparams.WatchParam{}
 
-	if svcs.conf.Blockchain.ChainProvider == blockchain.ProviderNullChain {
-		svcs.pow = pow.NewNoop()
-	} else {
-		pow := pow.New(svcs.log, svcs.conf.PoW, svcs.timeService)
-		svcs.pow = pow
-		svcs.snapshot.AddProviders(pow)
-		powWatchers = []netparams.WatchParam{
-			{
-				Param:   netparams.SpamPoWNumberOfPastBlocks,
-				Watcher: pow.UpdateSpamPoWNumberOfPastBlocks,
-			},
-			{
-				Param:   netparams.SpamPoWDifficulty,
-				Watcher: pow.UpdateSpamPoWDifficulty,
-			},
-			{
-				Param:   netparams.SpamPoWHashFunction,
-				Watcher: pow.UpdateSpamPoWHashFunction,
-			},
-			{
-				Param:   netparams.SpamPoWIncreasingDifficulty,
-				Watcher: pow.UpdateSpamPoWIncreasingDifficulty,
-			},
-			{
-				Param:   netparams.SpamPoWNumberOfTxPerBlock,
-				Watcher: pow.UpdateSpamPoWNumberOfTxPerBlock,
-			},
-			{
-				Param:   netparams.ValidatorsEpochLength,
-				Watcher: pow.OnEpochDurationChanged,
-			},
-		}
+	pow := pow.New(svcs.log, svcs.conf.PoW, svcs.timeService)
+	svcs.pow = pow
+	svcs.snapshot.AddProviders(pow)
+	powWatchers := []netparams.WatchParam{
+		{
+			Param:   netparams.SpamPoWNumberOfPastBlocks,
+			Watcher: pow.UpdateSpamPoWNumberOfPastBlocks,
+		},
+		{
+			Param:   netparams.SpamPoWDifficulty,
+			Watcher: pow.UpdateSpamPoWDifficulty,
+		},
+		{
+			Param:   netparams.SpamPoWHashFunction,
+			Watcher: pow.UpdateSpamPoWHashFunction,
+		},
+		{
+			Param:   netparams.SpamPoWIncreasingDifficulty,
+			Watcher: pow.UpdateSpamPoWIncreasingDifficulty,
+		},
+		{
+			Param:   netparams.SpamPoWNumberOfTxPerBlock,
+			Watcher: pow.UpdateSpamPoWNumberOfTxPerBlock,
+		},
+		{
+			Param:   netparams.ValidatorsEpochLength,
+			Watcher: pow.OnEpochDurationChanged,
+		},
 	}
```

Comment:
You can have this if like this:

```suggestion
	if svcs.conf.Blockchain.ChainProvider == blockchain.ProviderNullChain {
		svcs.pow = pow.NewNoop()
	} else {
		pow := pow.New(svcs.log, svcs.conf.PoW, svcs.timeService)
		svcs.pow = pow
		svcs.snapshot.AddProviders(pow)
	}
	
	powWatchers = []netparams.WatchParam{
			{
				Param:   netparams.SpamPoWNumberOfPastBlocks,
				Watcher: svcs.pow.UpdateSpamPoWNumberOfPastBlocks,
			},
			{
				Param:   netparams.SpamPoWDifficulty,
				Watcher: svcs.pow.UpdateSpamPoWDifficulty,
			},
			{
				Param:   netparams.SpamPoWHashFunction,
				Watcher: svcs.pow.UpdateSpamPoWHashFunction,
			},
			{
				Param:   netparams.SpamPoWIncreasingDifficulty,
				Watcher: svcs.pow.UpdateSpamPoWIncreasingDifficulty,
			},
			{
				Param:   netparams.SpamPoWNumberOfTxPerBlock,
				Watcher: svcs.pow.UpdateSpamPoWNumberOfTxPerBlock,
			},
			{
				Param:   netparams.ValidatorsEpochLength,
				Watcher: svcs.pow.OnEpochDurationChanged,
			},

```

Reply:
I have checked the way you suggested here and We discussed and this is not going to work because the `NoopEngine` returned by `pow.NewNoop` does not contain items like `UpdateSpamPoWNumberOfPastBlocks`, `UpdateSpamPoWDifficulty`, etc.